mne/preprocessing/_css.py

```diff
@@ -0,0 +1,100 @@
+# Author: John Samuelsson <johnsam@mit.edu>
+
+import numpy as np
+import mne
+from mne.utils import logger
+from numpy import linalg
+
+
+def temp_proj(ref_data_2, ref_data_1, raw_data, n_proj=6):
+    """
+    This temporal projection procedure removes the common signal subspace 
+    between ref_data_2 and ref_data_1 from raw_data using n_proj number of 
+    projection vectors. Normally used for cortical signal suppression, where
+    ref_data_1 is gradiometer data, ref_data_2 is magnetometer data and 
+    raw_data is EEG data.
+
+    Parameters
+    ----------
+    ref_data_2 : np.ndarray of float, shape (n_sensors_2, n_times)
+        The magnetometer data for CSS. Can use either all magnetometer data or
+        a few selected sensors close to a region to be suppressed.
+    ref_data_1 : np.ndarray of float, shape (n_sensors_1, n_times)
+        The gradiometer data for CSS. Can use either all gradiometer data or
+        a few selected sensors close to a region to be suppressed.
+    raw_data : np.ndarray of float, shape (n_sensors_raw, n_times)
+        The data to be filtered, typically the EEG data.
+    n_proj : int
+        The number of projection vectors.
+
+    Returns
+    -------
+    filtered_data : np.ndarray of float, shape (n_sensors_raw, n_times)
+        The filtered data.
+    """
+    # Orthonormalize gradiometer and magnetometer data by a QR decomposition
+    Qg = linalg.qr(ref_data_1.T)[0]
+    Qm = linalg.qr(ref_data_2.T)[0]
+
+    # Calculate cross-correlation
+    C = np.dot(Qg.T, Qm)
+
+    # Channel weights for common temporal subspace by SVD of cross-correlation
+    Y, S, Z = linalg.svd(C)
+
+    # Get temporal signals from channel weights
+    u = np.dot(Qg, Y)
+
+    # Project out common subspace
+    filtered_data = raw_data
+    for i in range(n_proj):
+        proj_vec = u[:, i].reshape(u.shape[0], 1)
+        weights = np.dot(filtered_data, proj_vec)
+        filtered_data = filtered_data - np.dot(weights, proj_vec.T)
+
+    return filtered_data
+
+
+def cortical_signal_suppression(evoked, n_proj=6, show_psd=False):
+    """Remove the cortical contribution to evoked by cortical signal
+    suppression (CSS).
+
+    This method removes the common signal subspace between the magnetometer
+    data and the gradiometer data from the EEG data. This is done by a temporal
+    projection using n_proj number of projection vectors.
+
+    Parameters
+    ----------
+    evoked : evoked object
+        The evoked object with the averaged epochs.
+    n_proj : int
+        The number of projection vectors.
+
+    Returns
+    -------
+    evoked_subcortical : instance of Evoked
+        The evoked object with filtered EEG data.
+    """
+    evoked_subcortical = evoked.copy()
```

Comment:
is the copy necessary?

Reply:
I think so, otherwise you will do an in-place change of evoked object right which I don't want (I want to return a new evoked with subcortical signal while keeping the original evoked as well)